### PR TITLE
std: page_size design and implementation

### PR DESCRIPTION
**Pull request by matu3ba**
Justification: Kernel compilation or configuration determines page size
from available page sizes of architecture (for optimal performance).
Currently, Zig is not portable to that regard.

This is WIP to get CI-coverage, ideas and clarify questions.

Follow-up will be large page abstraction API.

Depends on #16638.

### CODE REVIEW

File: lib/std/os.zig [hunk 1]
```diff
@@ -4337,6 +4338,47 @@ pub fn fork() ForkError!pid_t {
...
+            if (rc != .SUCCESS) {
+                return error.UnknownPageSize;
+            }
+            return sbi.PageSize;
+        },
+        // zig fmt: off
+        .dragonfly, .freebsd, .fuchsia, .haiku,
+        .macos, .ios, .watchos, .tvos,
+        .minix, .netbsd, .openbsd, .solaris, => {
+            return sysconf(std.c._SC.PAGESIZE) catch return error.UnknownPageSize;
```
Comment (kubkon) [inline lib/std/os.zig:4374]: Seeing how many syscalls this PR has introduced I think it'd be wise to at least try and cache and reuse the result of the first syscall.

Comment (kubkon) [reply lib/std/os.zig:4374]: Also, how come on macOS you use `std.c.sysconf` when a direct Mach call is available that does not require elevated privs (you can always inquire about yourself)?

Comment (matu3ba) [reply lib/std/os.zig:4374]: > Seeing how many syscalls this PR has introduced I think it'd be wise to at least try and cache and reuse the result of the first syscall.

At least on Linux I dont see a syscall done within libc and unless Mac is weird, it should do kinda the same. For Windows, I would need to have another look. Maybe I am missing something and strace is lying?
```
misterspoon@pc ~/d/g/z/z/learn_page_size (learn_page_size)> fd -t d -u "zig-out|zig-cache" -x rm -fr {};
misterspoon@pc ~/d/g/z/z/learn_page_size (learn_page_size)> ./deb/bin/zig test ./lib/std/os/test.zig --test-filter "getDefaultPageSize smoke test"  -freference-trace --test-no-exec -lc
misterspoon@pc ~/d/g/z/z/learn_page_size (learn_page_size)> fd . zig-cache/
zig-cache/h/
zig-cache/o/
zig-cache/o/592c12bccaabd14f03a3967b18c968c8/
zig-cache/o/592c12bccaabd14f03a3967b18c968c8/builtin.zig
zig-cache/o/592c12bccaabd14f03a3967b18c968c8/test
zig-cache/o/592c12bccaabd14f03a3967b18c968c8/test.o
zig-cache/z/
zig-cache/z/3bcfc54ac2dd103da5f8cfd3b6d67624
misterspoon@pc ~/d/g/z/z/learn_page_size (learn_page_size)> strace zig-cache/o/592c12bccaabd14f03a3967b18c968c8/test
execve("zig-cache/o/592c12bccaabd14f03a3967b18c968c8/test", ["zig-cache/o/592c12bccaabd14f03a3"...], 0x7fff82191f00 /* 83 vars */) = 0
brk(NULL)                               = 0x1dc9000
arch_prctl(0x3001 /* ARCH_??? */, 0x7ffd9d806990) = -1 EINVAL (Das Argument ist ungültig)
access("/etc/ld.so.preload", R_OK)      = 0
openat(AT_FDCWD, "/etc/ld.so.preload", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=0, ...}, AT_EMPTY_PATH) = 0
close(3)                                = 0
openat(AT_FDCWD, "/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 3
newfstatat(3, "", {st_mode=S_IFREG|0644, st_size=184619, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 184619, PROT_READ, MAP_PRIVATE, 3, 0) = 0x7ff0ab48e000
close(3)                                = 0
openat(AT_FDCWD, "/usr/lib/libc.so.6", O_RDONLY|O_CLOEXEC) = 3
read(3, "\177ELF\2\1\1\3\0\0\0\0\0\0\0\0\3\0>\0\1\0\0\0\220~\2\0\0\0\0\0"..., 832) = 832
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
newfstatat(3, "", {st_mode=S_IFREG|0755, st_size=2366624, ...}, AT_EMPTY_PATH) = 0
mmap(NULL, 8192, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ff0ab48c000
pread64(3, "\6\0\0\0\4\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0@\0\0\0\0\0\0\0"..., 784, 64) = 784
mmap(NULL, 2411920, PROT_READ, MAP_PRIVATE|MAP_DENYWRITE, 3, 0) = 0x7ff0ab200000
mmap(0x7ff0ab226000, 1437696, PROT_READ|PROT_EXEC, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x26000) = 0x7ff0ab226000
mmap(0x7ff0ab385000, 348160, PROT_READ, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x185000) = 0x7ff0ab385000
mmap(0x7ff0ab3da000, 417792, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_DENYWRITE, 3, 0x1d9000) = 0x7ff0ab3da000
mmap(0x7ff0ab440000, 52624, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_FIXED|MAP_ANONYMOUS, -1, 0) = 0x7ff0ab440000
close(3)                                = 0
mmap(NULL, 12288, PROT_READ|PROT_WRITE, MAP_PRIVATE|MAP_ANONYMOUS, -1, 0) = 0x7ff0ab489000
arch_prctl(ARCH_SET_FS, 0x7ff0ab489740) = 0
set_tid_address(0x7ff0ab489a10)         = 1660249
set_robust_list(0x7ff0ab489a20, 24)     = 0
rseq(0x7ff0ab48a060, 0x20, 0, 0x53053053) = 0
mprotect(0x7ff0ab3da000, 409600, PROT_READ) = 0
mprotect(0x2cb000, 12288, PROT_READ)    = 0
mprotect(0x7ff0ab4ed000, 8192, PROT_READ) = 0
prlimit64(0, RLIMIT_STACK, NULL, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
munmap(0x7ff0ab48e000, 184619)          = 0
prlimit64(0, RLIMIT_STACK, NULL, {rlim_cur=8192*1024, rlim_max=RLIM64_INFINITY}) = 0
prlimit64(0, RLIMIT_STACK, {rlim_cur=16384*1024, rlim_max=RLIM64_INFINITY}, NULL) = 0
rt_sigaction(SIGSEGV, {sa_handler=0x22ea40, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART|SA_RESETHAND|SA_SIGINFO|0xffffffff00000000, sa_restorer=0x7ff0ab23e710}, NULL, 8) = 0
rt_sigaction(SIGILL, {sa_handler=0x22ea40, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART|SA_RESETHAND|SA_SIGINFO|0xffffffff00000000, sa_restorer=0x7ff0ab23e710}, NULL, 8) = 0
rt_sigaction(SIGBUS, {sa_handler=0x22ea40, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART|SA_RESETHAND|SA_SIGINFO|0xffffffff00000000, sa_restorer=0x7ff0ab23e710}, NULL, 8) = 0
rt_sigaction(SIGFPE, {sa_handler=0x22ea40, sa_mask=[], sa_flags=SA_RESTORER|SA_RESTART|SA_RESETHAND|SA_SIGINFO|0xffffffff00000000, sa_restorer=0x7ff0ab23e710}, NULL, 8) = 0
rt_sigaction(SIGPIPE, {sa_handler=0x22ec10, sa_mask=[], sa_flags=SA_RESTORER, sa_restorer=0x7ff0ab23e710}, NULL, 8) = 0
ioctl(2, TCGETS, {c_iflag=ICRNL|IUTF8, c_oflag=NL0|CR0|TAB0|BS0|VT0|FF0|OPOST|ONLCR, c_cflag=B0|CS8|CREAD, c_lflag=ISIG|ICANON|ECHO|ECHOE|ECHOK|IEXTEN|ECHOCTL|ECHOKE, ...}) = 0
gettid()                                = 1660249
write(2, "Test [1/1] test.getDefaultPageSi"..., 49Test [1/1] test.getDefaultPageSize smoke test... ) = 49
)            = 9
write(2, "All ", 4All )                     = 4
write(2, "1", 11)                        = 1
write(2, " tests passed.\n", 15 tests passed.
)        = 15
exit_group(0)                           = ?
+++ exited with 0 +++
```

Getting test file from cache via `fd . zig-cache/`.

Comment (matu3ba) [reply lib/std/os.zig:4374]: > Also, how come on macOS you use std.c.sysconf when a direct Mach call is available that does not require elevated privs (you can always inquire about yourself)?

1. Can you point me to the darwin/mac docs, where it phrases sysconf need elevated privileges? Afaiu, sysconf only queries values to get system values and most of it is very basic stuff.
2. Ok, I found usage now in `ptraceAttach` `const mach_task = try std.os.darwin.machTaskForPid(pid);` and will use it instead.
3. I would expect basic stuff like reading what the loader has written into the darwin version of the aux-vector to work correct, but I suspect you indicate it is not.

**Comment by andrewrk** [comment]
Closing abandoned pull request (draft, all checks failing, conflicts, 30+ days old)

**Pull request closed without merging**